docs/contributing/unit-test-guideline.md

```diff
@@ -0,0 +1,206 @@
+
+If a function or method accept multiple inputs, we should try to create test cases to cover all the inputs.
+
+(Note: The below `calcSum()` is diffrent function from `calsSum(val ...int32)` which we mention before.)
+
+```go
+func calcAverageDiff(val1 []int32, val2 []int32) (diff float64) {
+	var ave1, ave2 float64
+	if len(val1) == 0 && len(val2) == 0 {
+		return
+	}
+	if len(val1) != 0 {
+		ave1 = float64(calcSum(val1)) / float64(len(val1))
+	}
+	if len(val2) != 0 {
+		ave2 = float64(calcSum(val2)) / float64(len(val2))
+	}
+
+	diff = math.Abs(ave1 - ave2)
+	return
+}
+```
+
+When `calcAverageDiff()` is given, the test patterns are below:
+
+| len(val1) | len(val2) | option |
+|:-----:|:-----:|:-----:|
+| 0  | 0  | - |
+| 0  | >0 | - |
+| >0 | 0  | - |
+| >0 | >0 | ave1 > ave2 |
+| >0 | >0 | ave1 < ave2 |
+| >0 | >0 | ave1 = ave2 |
+| >0 | >0 | ave1 = {math.MinInt32}, ave2 = {math.MaxInt32} |
+| >0 | >0 | ave1 = {math.MaxInt32, math.MaxInt32}, ave2 = {0} |
+
+At a glance, one of the last 2 patterns is enough, but these will help us to notice the bug in dependency.
+It will avoid the unexpected error due to update dependencies.
+
+You have to create unit tests for error patterns as the same as success patterns.
+
+#### Advanced
+
+##### Robust boudary test
+
+The previous section is about the basic test cases.
+To cover more test coverage, the (robust) boundary test should be applied.
+A boundary test is testing values on or near the boundaries of allowed inputs.
+
+```go
+func AgeTest(val int32) (ok bool) {
+	if val >= 20 && val < 65 {
+		ok = true
+	}
+	return
+}
+```
+
+When the above function is given as target, the minimum number of tests based on the "Robust Boundary" test is 7.
+  - the input `val` cases: 19, 20, 21, 30, 64, 65, 66
+  
+The robust boundary test requires 6N+1 test cases (N is the number of inputs for target function).
+Increasing the input arguments will also increase the number of required test cases.
+Sometimes, it requires a lot of resources to create/maintain test.
+So, you may test the critical cases using boundary value and have a sense of purpose.
+
+#### Equivalence Class Testing
+
+Equivalence Class Testing is grouping the values together such that all members of the group are considered the same and testing one value from each.
+```go
+func AgeTest(val int32) (ok bool) {
+	if val >= 20 && val < 65 {
+		ok = true
+	}
+	return
+}
+```
+
+When the above function is given as target, you have to create 3 groups and pick one value from each group.
+
+|Desc|Input range|Chose Input|
+|:-----:|:-----:|:-----:|
+| Under 20    | {0, ...,  19 }| 5  |
+| worker range| {20, ..., 64} | 29 |
+| Senior      | {65, ...    } | 72 |
+
+### Vald Style
+
+In the Vald, we create unit tests based on the basic test case.
+And, you may also create unit tests based on robust boundary tests or equivalence class tests if needed.
+
+But, we have to take care about that the Vald is developed using Go.
```

Comment:
```suggestion
But, we have to take care that the Vald is developed using Go.
```

Reply:
```suggestion
But, we have to take care of the Vald is developed using Go.
```

Reply:
thank you for your suggestion. 🙇
It looks good to me.
I think your suggestion is good.